onmt/inputters/inputter.py

```diff
@@ -220,12 +161,10 @@ def make_features(batch, side):
         data, lengths = batch.__dict__[side]
     else:
         data = batch.__dict__[side]
-        if side == 'src' and hasattr(batch, 'src_lengths'):
-            lengths = batch.src_lengths
-        else:
-            lengths = None
+        lengths = None
 
-    if isinstance(batch.__dict__[side], tuple) or side == 'tgt':
+    if not data.dim() == 4 and (
+            isinstance(batch.__dict__[side], tuple) or side == 'tgt'):
```

Comment:
not sure to follow this data.dim() == 4 test

Reply:
Oh yeah, that's not a good change. Sorry.

It used to be ``if data_type == 'text':``, but #1184 axed the data_type argument and checks if batch.src is a tuple since, before, only text was a tuple. Now audio is a tuple too.

I'd suggest going back to passing data_type into get_features now that there's an actual ambiguity. Is that okay?

Reply:
I don't think you have access to data_type in all the places where you call make_features

Reply:
Actually I'm looking at the usages and that's not okay. I wonder if there's a way to attach the data type to the batch.

Reply:
In the long run, by the time it's batched it shouldn't be necessary to know the data type. It's just a batch of data that the model should be able to handle. It only comes up right now because our way of handling feature embeddings (a feature almost no one uses) is super hacky.

Reply:
Well what would be a better way ? When I used to use the Lua version, it was very convenient to use feature embeddings and somehow efficient. In onmt-py today just the fact that we don't support target features makes this useless but it would be a good thing to have. no ?

Reply:
Feature embeddings are undoubtedly a good thing to have that I have used in my own research. The question is just when, where, and how they get numericalized. I think the best way would be some kind of multilevel field that takes an input which is words plus arbitrarily many features and when numericalized produced the sort of stacked tensor that the embeddings module expects. The `NestedField` in torchtext almost, but not quite, fits the bill.

It would save the trainer and translator from having to reason about the types of the source and target data.

Reply:
oh ok I misread the "(almost no one uses)" comment.
So in the meantime, are we ok to set a data_type attribute to batch ?

Reply:
Yes, provided we understand that it's a temporary solution until such a time as the data processing pipeline is capable of numericalizing its own input.